### DISCUSSION TITLE
Valis Optomisation and batch processing

### DISCUSSION THREAD

**Discussion by Justin-Krieg** [MathOnco/valis#62] [Q&A] [answered]
Hello, this discussion is regarding optimising VALIS, and I had a couple of questions: 
Is there an optimal input file format for VALIS? Currently I've been converting my multiscene czi files to ome-tif as some of the scenes are in the wrong order. However I was curious if there was an optimal input format that would allow for the quickest registration?

Adding to this, the intention was to get VALIS to register a batch of slides. I was considering having a folder for each slide with each round of CyCIF in it and thenwrite a for loop in python. However I'm using the docker environment which requires an input and output directory which would change for each slide. Looping through docker instances feels rather inefficient. 

Cheers,

Justin

**Comment by cdgatenbee** [comment, marked as answer]
Hi @kriegy98,
Valis relies heavily on [libvips](https://github.com/libvips/libvips), which supports quite a few formats, including WSI formats that can be read by [OpenSlide](https://openslide.org/) (libvips does support ome-tiff, even though it's not on the list). As such, any of the formats compatible with libvips will be read in very quickly. Unfortunately, czi isn't one of those formats. In that case, valis has to build the libvips Image by reading it in tiles via Bioformats, which unfortunately takes a bit longer to execute.

Regarding your 2nd question, does the code in https://github.com/MathOnco/valis/discussions/61 help? The docker container should just launch a Python interpreter, and so I think you could use that code (or something like it) to register your CyCIF rounds. If that won't work, let me know and we can try to figure out how to accomplish your goal.

Best,
-Chandler

**Reply by Justin-Krieg** [reply]
Hi @cdgatenbee 

Thank you for your information regarding different supported formats, i think i'll keep the pipeline where i convert files from czi to ome.tiff using zen blue, organise scenes and then run it from there. 

Regarding point 2, i think i was confused from the documentation where the docker container needed the src_dir and dst_dir as arguments. I now realise that those parameters can just be read from the python file. So i ended up creating a for loop iterating through subdirectories with images that needed to be aligned and merged. Again putting this here for anyone else who needs to do this. 
```
from valis import registration, valtils
import os
rootdir = '/var/tmp/Input'
for entry in os.listdir(rootdir):
    if not os.path.isdir(entry):
        input = os.path.join(rootdir, entry)
        output_dir = os.path.join(rootdir, 'Processing')
        output = os.path.join(rootdir, entry + ".ome.tif")
        os.makedirs(output_dir)
        registrar = registration.Valis(input, output_dir, max_processed_image_dim_px = 1500, max_non_rigid_registration_dim_px = 2000)
        rigid_registrar, non_rigid_registrar, error_df = registrar.register()
        registrar.warp_and_merge_slides(output, drop_duplicates=True)
registration.kill_jvm()
```
And then i ran this using this docker command:  `docker run --memory=55g -v "C:/Slides/:/var/tmp/" cdgatenbee/valis-wsi:latest python3 /var/tmp/register.py`
